polygeist/include/mlir/Dialect/Polygeist/Analysis/SYCLIDAndRangeAnalysis.h

```diff
@@ -0,0 +1,83 @@
+//===- SYCLIDAndRangeAnalysis.h - Analysis for sycl::id/range ---*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file contains an analysis to derive interesting properties about
+// sycl::id and sycl::range in SYCL host code.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MLIR_DIALECT_POLYGEIST_ANALYSIS_SYCLIDANDRANGEANALYSIS_H
+#define MLIR_DIALECT_POLYGEIST_ANALYSIS_SYCLIDANDRANGEANALYSIS_H
+
+#include "mlir/Analysis/DataFlowFramework.h"
+#include "mlir/Pass/AnalysisManager.h"
+
+namespace mlir {
+namespace polygeist {
+
+/// Represents information about a `sycl::id` or `sycl::range` gathered from its
+/// construction.
+class IDRangeInformation {
+public:
+  IDRangeInformation();
+
+  explicit IDRangeInformation(size_t dim);
+
+  explicit IDRangeInformation(llvm::ArrayRef<size_t> constVals);
+
+  /// Returns true if the id/range is always constructed with the same number of
+  /// dimensions.
+  bool hasFixedDimensions() const;
+
+  /// Returns the number of dimensions for an id or range, in case it is always
+  /// constructed with the same number of dimensions.
+  size_t getNumDimensions() const;
+
+  /// Returns true if the id/range is always constructed with the same constant
+  /// values.
+  bool isConstant() const;
+
+  /// Returns the constant values with which this id/range is constructed, in
+  /// case it is always constructed with the same constant values.
+  const llvm::SmallVector<size_t, 3> &getConstantValues() const;
+
+  const IDRangeInformation join(const IDRangeInformation &other) const;
+
+private:
+  std::optional<size_t> dimensions;
+
+  std::optional<llvm::SmallVector<size_t, 3>> constantValues;
+
+  friend raw_ostream &operator<<(raw_ostream &, const IDRangeInformation &);
+};
+
+/// Analysis to determine properties of interest about `sycl::id` or
+/// `sycl::range` from their construction.
+class SYCLIDAndRangeAnalysis {
+public:
+  SYCLIDAndRangeAnalysis(Operation *op, AnalysisManager &am);
+
+  void initialize(bool useRelaxedAliasing = false);
```

Comment:
Can we remove the default (have the consumer explicitly pass whether to use relaxed aliasing) and also return a reference to this class, like:
```
/// Consumers of the analysis must call this member function immediately after construction.
SYCLIDAndRangeAnalysis& initialize(bool useRelaxedAliasing);
```
That way the consumer can construct and initialize the analysis in one line:

```
 auto &IDRangeAnalysis = getAnalysis<polygeist::SYCLIDAndRangeAnalysis>().initialize(relaxedAliasing);

Reply:
Done.